presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/SimplifyCoalesceExpressionOverJoin.java

```diff
@@ -0,0 +1,97 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.JoinNode.EquiJoinClause;
+import io.prestosql.sql.planner.plan.JoinNode.Type;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.tree.CoalesceExpression;
+import io.prestosql.sql.tree.Expression;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.FULL;
+import static io.prestosql.sql.planner.plan.Patterns.join;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+
+public class SimplifyCoalesceExpressionOverJoin
+        implements Rule<ProjectNode>
+{
+    private static final Capture<JoinNode> CHILD = newCapture();
+
+    private static final Pattern<ProjectNode> PATTERN = project()
+            .matching(projectNode -> projectNode.getAssignments().getExpressions().stream()
+                    .anyMatch(expression -> expression instanceof CoalesceExpression))
+            .with(source().matching(
+                    join().capturedAs(CHILD).matching(
+                            joinNode -> !joinNode.getType().equals(FULL))));
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(ProjectNode node, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(CHILD);
+        ImmutableMap.Builder<Set<Expression>, Expression> builder = ImmutableMap.builder();
+
+        Function<EquiJoinClause, Expression> joinClauseMapper = getEquiJoinClauseMapper(joinNode.getType());
+
+        for (JoinNode.EquiJoinClause joinClause : joinNode.getCriteria()) {
+            builder.put(ImmutableSet.of(joinClause.getLeft().toSymbolReference(), joinClause.getRight().toSymbolReference()), joinClauseMapper.apply(joinClause));
+        }
+        Map<Set<Expression>, Expression> coalesceExpressionMapper = builder.build();
+
+        Assignments translatedAssignments = node.getAssignments().rewrite(expression -> {
+            if (expression instanceof CoalesceExpression) {
+                CoalesceExpression coalesceExpression = (CoalesceExpression) expression;
+                return coalesceExpressionMapper.getOrDefault(ImmutableSet.copyOf(coalesceExpression.getOperands()), coalesceExpression);
+            }
```

Comment:
We could generalize this. The key insight is that with in an equi-join the keys need to be equal or the outer one has to be non-null, so any instance of the inner key can be removed from the expression, and if we end up with a single element (i.e., the outer key), we can remove the coalesce call altogether. So, for instance:

```
coalesce(outerKey, innerKey, 1) ==> coalesce(outerKey, 1)

coalesce(outerKey, innerKey, outerKey, innerKey) => outerKey

coalesce(innerKey, outerKey, innerKey) => outerKey
```

Reply:
> with in an equi-join the keys need to be equal or the outer one has to be non-null

Did you mean "keys need to be equal or the inner one has to be null" ?

Reply:
Almost, I actually meant to say "with in an equi-join the keys need to be equal, or the outer one has to be non-null *and the inner one null*"

Reply:
The outer one can still be null

Reply:
Ah, yes, if the source row has nulls.